### DISCUSSION TITLE
SetPickupToDeliveryLimitFunctionForPair

### DISCUSSION THREAD

**Discussion by lewis-xuwei** [google/or-tools#3362] [Routing (and legacy CP) questions] [answered]
SetPickupToDeliveryLimitFunctionForPair

failed using it to restrict the total distance between a pair of pickup and delivery

```py
def pick_del_pair_total_distance(from_index, to_index):
        from_node = from_index
        to_node = to_index
        return data["distance_matrix"][from_node][to_node] * 6

    # Add Distance constraint.
    dimension_name = "Distance"
    routing.AddDimension(
        transit_callback_index,
        0,  # no slack
        3000,  # vehicle maximum travel distance
        True,  # start cumul to zero
        dimension_name,
    )
    distance_dimension = routing.GetDimensionOrDie(dimension_name)
    distance_dimension.SetGlobalSpanCostCoefficient(100)

    # Define Transportation Requests.
    i = 0
    for request in data["pickups_deliveries"]:
        pickup_index = manager.NodeToIndex(request[0])
        delivery_index = manager.NodeToIndex(request[1])
        routing.AddPickupAndDelivery(pickup_index, delivery_index)
        routing.solver().Add(
            routing.VehicleVar(pickup_index) == routing.VehicleVar(delivery_index)
        )
        routing.solver().Add(
            distance_dimension.CumulVar(pickup_index)
            <= distance_dimension.CumulVar(delivery_index)
        )
    for i in range(len(data["pickups_deliveries"])):
        print(i)
        distance_dimension.SetPickupToDeliveryLimitFunctionForPair(pick_del_pair_total_distance, i)
```

**Comment by lperron** [comment, marked as answer]
what does it means ?

**Reply by lewis-xuwei** [reply]
My purpose is to restrict the actual route length between pickup and delivery pairs (PD pairs).

Given a route $$[0, \cdots, p_i, \cdots, d_i,\cdots,  0]$$, thePD pairs are $$[p_i, d_i]$$. I want to add a constraint as follows: the length of subroute of pair $[p_i, d_i]$ cannot exceeds the straight distance between $p_i$ and $d_i$

For example, a route is [0,1, 2, 3, 4, 0], the pickup and delivery pairs [[1,3], [2,4]]. Thus the  actual route length of PD pairs [1, 3] is length of sub-route [1,2,3] ($l_{[1,2,3]}$), and the straight distance of [1,3] is $d_{1,3}$. So my constraint is: l_{[1,2,3]} \leqslant \alpha \times d_{1,3}, where the \alpha is call nonlinear coefficent.

**Reply by lewis-xuwei** [reply]
wait for your excellent solution

**Reply by Mizux** [reply]
Something along this line ?
```python
solver = routing.solver()
distance_dimension = routing.GetDimensionOrDie('Distance')

pickup_index = manager.NodeToIndex(1)
delivery_index = manager.NodeToIndex(3)
solver.Add(distance_dimension.CumulVar(delivery_index) - distance_dimension.CumulVar(pickup_index) <=
  distance[{1, 3}] * coef)
```

**Reply by lewis-xuwei** [reply]
Taking the official example that you can refer at https://developers.google.cn/optimization/routing/pickup_delivery#complete_programs as the case, I add the above solution in it as follows.

```
for request in data["pickups_deliveries"]:
        pickup_index = manager.NodeToIndex(request[0])
        delivery_index = manager.NodeToIndex(request[1])
        routing.AddPickupAndDelivery(pickup_index, delivery_index)
        routing.solver().Add(
            routing.VehicleVar(pickup_index) == routing.VehicleVar(delivery_index)
        )
        routing.solver().Add(
            distance_dimension.CumulVar(pickup_index)
            <= distance_dimension.CumulVar(delivery_index)
        )
        ############################################################
        routing.solver().Add(
            distance_dimension.CumulVar(delivery_index)
            - distance_dimension.CumulVar(pickup_index)
            <= data["distance_matrix"][request[0]][request[1]] * 1.5 # coeff
        )
        ############################################################
```

The fact shows it does not work, and throws an fatal error.

```
Traceback (most recent call last):
  File "c:/Users/Administrator/Desktop/play-ground/or-tools-routing.py", line 472, in <module>
    main()
  File "c:/Users/Administrator/Desktop/play-ground/or-tools-routing.py", line 429, in main
    distance_dimension.CumulVar(delivery_index)
TypeError: '<=' not supported between instances of 'IntExpr' and 'float'
```

**Comment by lperron** [comment]
The solver does not accept floating point values.

**Reply by lewis-xuwei** [reply]
Thank you with my biggest respect. It works by using the following code!

```py
routing.solver().Add(
            distance_dimension.CumulVar(delivery_index)
            - distance_dimension.CumulVar(pickup_index)
            <= int(data["distance_matrix"][request[0]][request[1]] * 1.5) + 1
        )
```

Is right?

The reason is that C++ code didn't support the operation between IntExpr and float number.

**Reply by Mizux** [reply]
Maybe this code is the entry point of what you may looking for...

https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/python/constraint_solver.i#L346-L347
...
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/python/constraint_solver.i#L406-L423

ps: https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/ortools/constraint_solver/expr_cst.cc#L267-L287

**Comment by lperron** [comment]
I understand what it does. I do not understand the issue:
  - it does not exist ?
  - it does not work ?
  - there are not examples ?

Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
